### PR TITLE
Feature/error reporting

### PR DESCRIPTION
**Pull request by petersteneteg**
Changes proposed in this PR:
 * The indicator light on processors will now indicate errors for each processor

image

Is ready has a reason
image

source processor error:
image

### CODE REVIEW

File: include/inviwo/qt/editor/processorgraphicsitem.h [hunk 1]
```diff
@@ -119,18 +124,23 @@ class IVW_QTEDITOR_API ProcessorGraphicsItem : public EditorGraphicsItem,
...
-#if IVW_PROFILING
     virtual void onProcessorAboutToProcess(Processor*) override;
     virtual void onProcessorFinishedProcess(Processor*) override;
-#endif
+
+    virtual void onProcessorStartBackgroundWork(Processor*, size_t jobs) override {
+        backgroundJobs_ += jobs;
+    };
+    virtual void onProcessorFinishBackgroundWork(Processor*, size_t jobs) override {
+        backgroundJobs_ -= jobs;
```
Comment (martinfalk) [inline include/inviwo/qt/editor/processorgraphicsitem.h:134]: overflow/underflow handling due to size_t?

**Review by martinfalk — DISMISSED** [review]

File: include/inviwo/core/processors/processorstatus.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#include <inviwo/core/util/sourcecontext.h>
+#include <inviwo/core/util/exception.h>
+
+#include <string_view>
+
+namespace inviwo {
+
+class IVW_CORE_API ProcessorStatus {
+public:
+    enum class Status { Ready, NotReady, Error };
```
Comment (ResearchDaniel) [inline include/inviwo/core/processors/processorstatus.h:42]: Invalid is not necessary?

File: modules/vectorfieldvisualizationgl/src/processors/3d/streamparticles.cpp [hunk 3]
```diff
@@ -169,6 +182,15 @@ StreamParticles::StreamParticles(InviwoApplication* app)
 }
 
 void StreamParticles::initializeResources() {
+    if (OpenGLCapabilities::getOpenGLVersion() < 430) {
+        isReady_.setUpdate([]() -> ProcessorStatus {
+            static constexpr std::string_view error{
+                "The StreamParticles processor requires OpenGL 4.3"};
+            return {ProcessorStatus::Error, error};
+        });
+        throw Exception(IVW_CONTEXT, "The StreamParticles processor requires OpenGL 4.3");
```
Comment (ResearchDaniel) [inline modules/vectorfieldvisualizationgl/src/processors/3d/streamparticles.cpp:191]: Could it be that it is always desirable to set isReady with error status when an exception is thrown? In that case, could we manage this behind the scenes? Or make a new type of exception that does it for you?

File: modules/vectorfieldvisualizationgl/src/vectorfieldvisualizationglmodule.cpp [hunk 4]
```diff
@@ -80,10 +78,7 @@ VectorFieldVisualizationGLModule::VectorFieldVisualizationGLModule(InviwoApplica
     registerProcessor<TMIP>();
     registerProcessor<VectorFieldGenerator4D>();
 
-    auto& gl = app->getModuleByType<OpenGLModule>()->getOpenGLCapabilities();
-    if (gl.getOpenGLVersion() >= 430) {
-        registerProcessor<StreamParticles>();
-    }
+    registerProcessor<StreamParticles>();
```
Comment (ResearchDaniel) [inline modules/vectorfieldvisualizationgl/src/vectorfieldvisualizationglmodule.cpp:81]: Nice!

File: include/inviwo/qt/editor/processorgraphicsitem.h (hunk 1 above)
Comment (petersteneteg) [reply include/inviwo/qt/editor/processorgraphicsitem.h:134]: As long as nothing else is wrong, that should never happen...

**Review by ResearchDaniel — APPROVED** [review]
Nice!
Would probably be nice to have a close button in the top right corner of the info.

File: modules/vectorfieldvisualizationgl/src/processors/3d/streamparticles.cpp (hunk 3 above)
Comment (petersteneteg) [reply modules/vectorfieldvisualizationgl/src/processors/3d/streamparticles.cpp:191]: Exceptions are also handled automatically

File: include/inviwo/core/processors/processorstatus.h (hunk 2 above)
Comment (petersteneteg) [reply include/inviwo/core/processors/processorstatus.h:42]: No, we have never had an invalid state for the isReady that is orthogonal

**Comment by petersteneteg** [comment]
> Nice! Would probably be nice to have a close button in the top right corner of the info.

If you unselect the processor it is gone

File: modules/vectorfieldvisualizationgl/src/processors/3d/streamparticles.cpp (hunk 3 above)
Comment (petersteneteg) [reply modules/vectorfieldvisualizationgl/src/processors/3d/streamparticles.cpp:191]: In a different way using the evaluator error handle

**Pull request merged**